server/engine/src/main/java/org/eclipse/lsp/cobol/core/engine/symbols/SymbolAccumulator.java

```diff
@@ -147,11 +148,35 @@ public Optional<SyntaxError> registerCodeBlockUsage(
     CodeBlockDefinitionNode definition = definitions.get(0);
     definition.addUsage(node.getLocality());
 
-    Optional.ofNullable(symbolTable.getParagraphMap().get(node.getName()))
-        .ifPresent(it -> it.addUsage(node.getLocality().toLocation()));
-    Optional.ofNullable(symbolTable.getSectionMap().get(node.getName()))
-        .ifPresent(it -> it.addUsage(node.getLocality().toLocation()));
+    Location location = node.getLocality().toLocation();
+    String definitionName = definition.getName();
 
+    // Try to resolve exact procedure identification
+    // 1. check if we are in a section
+    String trySameSection = definition.getNearestParentByType(NodeType.PROCEDURE_SECTION)
+            .map(n -> ((ProcedureSectionNode) n).getName())
+            .orElse(null);
+    // If there is paragraph with the same name - call it
+    if (trySameSection != null) {
+      ProcedureId defProcId = new ProcedureId(trySameSection, definitionName);
+      if (symbolTable.getProcedures().containsKey(defProcId)) {
+        symbolTable.getProcedures().get(defProcId).addUsage(location);
+        return Optional.empty();
+      }
+    }
+
+    // FIXME: if else.. go to paragraph/section whatever needs to be clarified
+    ProcedureId trySection = new ProcedureId(definitionName, null);
+    if (symbolTable.getProcedures().containsKey(trySection)) {
+      symbolTable.getProcedures().get(trySection).addUsage(location);
+      return Optional.empty();
+    }
+    for (Map.Entry<ProcedureId, CodeBlockReference> en: symbolTable.getProcedures().entrySet()) {
+      if (Objects.equals(definitionName, en.getKey().getParagraphName())) {
+        en.getValue().addUsage(location);
+        return Optional.empty();
+      }
+    }
```

Comment:
Is this the case of uniquely named paragraph, but defined in a different section?

Reply:
No, it's a pick a paragraph logic. I was going to implement resolution logic in the next pr. I don't know how it does it exactly work yet.
For example I'm not sure if PERFORM A. will perform paragraph or section it this case:
```
A SECTION.
B SECTION.
A.
```
and a couple of similar cases.